### PR TITLE
Fix predefined fields in selfservice

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #10884

### CODE REVIEW

**Review by orthagh — COMMENTED** [review]
as the code seems the same as `showForm` method, may we can add a new one with `$options` parameter passed as reference and returning `$predefined_fields` ?

This may be more complex to read, but i'm not very comfortable with duplicated code like this

File: src/Ticket.php [hunk 1]
```diff
@@ -4052,6 +4052,64 @@ public static function computeTco(CommonDBTM $item)
         return $totalcost;
     }
 
+    /**
+     * Return an array of predefined fields while also loading them into the $options array
+     *
+     * @param ITILTemplate $tt The ticket template to use
+     * @param array $options The current options array
```
Comment (orthagh) [inline src/Ticket.php:4059]: maybe add an indication here in comment to let developpers know `$options` is PASSED BY REFERENCE (in majuscule)

File: src/Ticket.php [hunk 2]
```diff
@@ -4052,6 +4052,64 @@ public static function computeTco(CommonDBTM $item)
...
+                        || (isset($options['_predefined_fields'][$predeffield])
+                            && ($options[$predeffield] == $options['_predefined_fields'][$predeffield]))
+                        || (isset($options[$key])
+                            && ($options[$key] != $tt->getID()))
+                    ) {
+                        $options[$predeffield]            = $predefvalue;
+                        $predefined_fields[$predeffield] = $predefvalue;
+                    }
+                } else { // Not defined options set as hidden field
+                    echo "<input type='hidden' name='$predeffield' value='$predefvalue'>";
```
Comment (cedric-anne) [inline src/Ticket.php:4095]: Could this HTML code be generated in twig template ?

**Review by orthagh — COMMENTED** [review]
One last thing, at the end of the method, keys added to $options may be the same as the one returned by the new method. maybe we can skip the reference parameter and merge the return into option ?

i didn't check live so i'm not sure, could you check that assertion ?

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by trasher** [comment]
please rebase to solve conflict

**Comment by cconard96** [comment]
> please rebase to solve conflict

Done

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**